### PR TITLE
Update torch version transformer and bert with Lazyapi

### PR DESCRIPTION
**Pull request by cheungdaven**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by cheungdaven** [comment]
@d2l-bot rebuild

File: chapter_computer-vision/image-augmentation.md [hunk 1]
```diff
@@ -359,12 +359,17 @@ def train_ch13(net, train_iter, test_iter, loss, trainer, num_epochs,
...
+               devices=d2l.try_all_gpus(), lazy=False):
+    """Train a model with mutiple GPUs (defined in Chapter 13).
+
+    Defined in :numref:`sec_image_augmentation`"""
     timer, num_batches = d2l.Timer(), len(train_iter)
     animator = d2l.Animator(xlabel='epoch', xlim=[1, num_epochs], ylim=[0, 1],
                             legend=['train loss', 'train acc', 'test acc'])
-    net = nn.DataParallel(net, device_ids=devices).to(devices[0])
+    if lazy:
+        net = net.to(devices[0])
```
Comment (astonzhang) [inline chapter_computer-vision/image-augmentation.md:370]: We should still support multi-GPU training when using Lazy APIs if possible. 
@AnirudhDagar Do you have ideas on this? Besides, did you remember why we have `nn.DataParallel(net, device_ids=devices).to(devices[0])`? Since we have `net = nn.DataParallel(net, device_ids=devices)` in http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_computational-performance/multiple-gpus-concise.html ,  is there any special consideration for using `to(devices[0])`?

File: chapter_natural-language-processing-pretraining/bert-pretraining.md [hunk 2]
```diff
@@ -175,7 +173,7 @@ def train_bert(train_iter, net, loss, vocab_size, devices, num_steps):
 ```{.python .input}
 #@tab pytorch
 def train_bert(train_iter, net, loss, vocab_size, devices, num_steps):
-    net = nn.DataParallel(net, device_ids=devices).to(devices[0])
+    net = net.to(devices[0])
```
Comment (astonzhang) [inline chapter_natural-language-processing-pretraining/bert-pretraining.md:176]: Ditto

File: chapter_computer-vision/image-augmentation.md [hunk 3]
```diff
@@ -359,12 +359,17 @@ def train_ch13(net, train_iter, test_iter, loss, trainer, num_epochs,
 #@tab pytorch
 #@save
 def train_ch13(net, train_iter, test_iter, loss, trainer, num_epochs,
-               devices=d2l.try_all_gpus()):
-    """Train a model with mutiple GPUs (defined in Chapter 13)."""
+               devices=d2l.try_all_gpus(), lazy=False):
```
Comment (astonzhang) [inline chapter_computer-vision/image-augmentation.md:362]: Let's try avoiding the extra `lazy` arg if possible.

File: chapter_natural-language-processing-applications/natural-language-inference-bert.md [hunk 4]
```diff
@@ -372,7 +370,8 @@ d2l.train_ch13(net, train_iter, test_iter, loss, trainer, num_epochs, devices,
 lr, num_epochs = 1e-4, 5
 trainer = torch.optim.Adam(net.parameters(), lr=lr)
 loss = nn.CrossEntropyLoss(reduction='none')
-d2l.train_ch13(net, train_iter, test_iter, loss, trainer, num_epochs, devices)
+d2l.train_ch13(net, train_iter, test_iter, loss, trainer, num_epochs, devices,
+               lazy=True)
```
Comment (astonzhang) [inline chapter_natural-language-processing-applications/natural-language-inference-bert.md:374]: Avoid `lazy` if possible

File: chapter_natural-language-processing-pretraining/bert-pretraining.md [hunk 5]
```diff
@@ -198,6 +196,7 @@ def train_bert(train_iter, net, loss, vocab_size, devices, num_steps):
             mlm_l, nsp_l, l = _get_batch_loss_bert(
                 net, loss, vocab_size, tokens_X, segments_X, valid_lens_x,
                 pred_positions_X, mlm_weights_X, mlm_Y, nsp_y)
+            #nn.DataParallel(net, device_ids=devices)
```
Comment (astonzhang) [inline chapter_natural-language-processing-pretraining/bert-pretraining.md:199]: Please remove this

File: d2l/torch.py [hunk 6]
```diff
@@ -991,11 +991,11 @@ class AdditiveAttention(nn.Module):
     """Additive attention.
 
     Defined in :numref:`sec_attention-scoring-functions`"""
-    def __init__(self, key_size, query_size, num_hiddens, dropout, **kwargs):
+    def __init__(self, num_hiddens, dropout, **kwargs):
```
Comment (astonzhang) [inline d2l/torch.py:994]: Please rerun `d2lbook build lib` after further code changes.

**Review by astonzhang — CHANGES_REQUESTED** [review]
Thanks! The code looks good to me except for those comments.

File: chapter_natural-language-processing-applications/natural-language-inference-bert.md [hunk 7]
```diff
@@ -313,7 +311,12 @@ class BERTClassifier(nn.Module):
         super(BERTClassifier, self).__init__()
         self.encoder = bert.encoder
         self.hidden = bert.hidden
-        self.output = nn.Linear(256, 3)
+        self.output = nn.LazyLinear(3)
+        self.encoder(torch.zeros([1,128], dtype=torch.long), 
+                     torch.zeros([1,128], dtype=torch.long),
+                     torch.zeros([1], dtype=torch.long))
+        self.hidden(torch.zeros([1,256]))
+        self.output(torch.zeros([1,256]))
```
Comment (astonzhang) [inline chapter_natural-language-processing-applications/natural-language-inference-bert.md:319]: Thanks! Are these workarounds best practice or necessary? Is it possible to avoid hardcoding 128 and 256?

File: chapter_natural-language-processing-pretraining/bert.md [hunk 8]
```diff
@@ -562,20 +560,24 @@ class BERTModel(nn.Block):
...
+        self.mlm = MaskLM(vocab_size, num_hiddens)
+        self.nsp = NextSentencePred()
+        
+        self.encoder(torch.zeros([1, 64], dtype=torch.long), 
+                    torch.zeros([1, 64], dtype=torch.long),
+                    torch.zeros([1], dtype=torch.long))
+        self.hidden(torch.zeros([1, num_hiddens]))
+        self.mlm(torch.zeros([1, 64, num_hiddens]), 
+                 torch.zeros([1, 10], dtype=torch.long))
+        self.nsp(torch.zeros([1, num_hiddens]))
```
Comment (astonzhang) [inline chapter_natural-language-processing-pretraining/bert.md:580]: Ditto

File: chapter_natural-language-processing-applications/natural-language-inference-bert.md (hunk 7 above)
Comment (astonzhang) [reply chapter_natural-language-processing-applications/natural-language-inference-bert.md:319]: Besides, please add space after commas

File: chapter_computer-vision/image-augmentation.md (hunk 1 above)
Comment (AnirudhDagar) [reply chapter_computer-vision/image-augmentation.md]: Q1.
To answer your first question we need to understand how `nn.DataParallel` works fundamentally.

When using `nn.DataParallel`, data should be initially stored on one GPU (the “primary GPU”). This also calls for the model to be moved to the GPU as the parameters need to be on the device for the forward pass. If there are no parameters initialized in the Pytorch layer (Lazy implementation) then it is technically not possible to copy those parameters when the dataparallel runs, essentially its four steps internally: `replicate`, `scatter`, `gather`, `parallel_apply`. Hence I don't think there is a direct way to use `nn.DataParallel` with Lazy implementation without a dry run of input dummy data into the model.

As mentioned in the docs, the expected way to use lazy API is to have a dry run, but in our case, this might lead to slightly not so good looking code:
> After construction, networks with lazy modules should first be converted to the desired dtype and placed on the expected device. This is because lazy modules only perform shape inference so the usual dtype and device placement behavior applies. The lazy modules should then perform “dry runs” to initialize all the components in the module. These “dry runs” send inputs of the correct size, dtype, and device through the network and to each one of its lazy modules. After this the network can be used as usual.

Q2.

> Besides, did you remember why we have nn.DataParallel(net, device_ids=devices).to(devices[0])? Since we have net = nn.DataParallel(net, device_ids=devices) in http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_computational-performance/multiple-gpus-concise.html , is there any special consideration for using to(devices[0])?

This is a very good question, in the [multiple gpus concise chapter](https://d2l.ai/chapter_computational-performance/multiple-gpus-concise.html#summary) we were able to use `net = nn.DataParallel(net, device_ids=devices)` without `net = nn.DataParallel(net, device_ids=devices).to(devices[0])`  because after the first training run (running with single GPU) the network params are already on GPU. When passing a single GPU dataparallel can move the model to that device. And then we don't need to explicitly move them to device[0] when we call training (run with two GPUs).

To be more precise if you skip the first training call, then the second function in that notebook should fail complaining about module parameters not residing on device[0]. This can be fixed if we also change `net = nn.DataParallel(net, device_ids=devices)` to `net = nn.DataParallel(net, device_ids=devices).to(devices[0])`

Q3.
Why `device[0]`?
There is no special reason, if we were to call only the dataparallel with 2 gpus (i.e the second train call with num_gpus=2) then we can move the parameters to any device, for that matter, `nn.DataParallel(net, device_ids=devices).to(devices[1])` should also work as long as the input tensors are also moved to Device 1.
But, we use devices[0] to make sure both functions work (because when calling train with `num_gpus=1` since there will be only one device 0 available), as a simple fix we write code in a fashion that is compatible to the scenario when only one device is available i.e (the first train call with num_gpus=1).


I hope this explanation clears the questions.

Comment (astonzhang) [reply chapter_computer-vision/image-augmentation.md]: Thanks for the detailed explanations!

Comment (cheungdaven) [reply chapter_computer-vision/image-augmentation.md]: Thanks for the explanations! I still have a question here. Will the training process run on a single GPU or multiple GPUs after we explicitly put dataparallel to devices[0]?

File: chapter_natural-language-processing-pretraining/bert-pretraining.md (hunk 2 above)
Comment (cheungdaven) [reply chapter_natural-language-processing-pretraining/bert-pretraining.md]: resolved.

File: chapter_computer-vision/image-augmentation.md (hunk 3 above)
Comment (cheungdaven) [reply chapter_computer-vision/image-augmentation.md]: Thanks for the suggestion. I have removed this argument.

File: chapter_natural-language-processing-applications/natural-language-inference-bert.md (hunk 4 above)
Comment (cheungdaven) [reply chapter_natural-language-processing-applications/natural-language-inference-bert.md]: Thanks. I have removed it.

File: chapter_natural-language-processing-pretraining/bert-pretraining.md (hunk 5 above)
Comment (cheungdaven) [reply chapter_natural-language-processing-pretraining/bert-pretraining.md]: Thanks for pointing out. it is removed in the latest submission.

File: d2l/torch.py (hunk 6 above)
Comment (cheungdaven) [reply d2l/torch.py:994]: It is done in the latest submission.

File: chapter_natural-language-processing-applications/natural-language-inference-bert.md (hunk 7 above)
Comment (cheungdaven) [reply chapter_natural-language-processing-applications/natural-language-inference-bert.md]: I removed these hardcodes and ran the forward pass with train_iter for lazy initialization.

File: chapter_natural-language-processing-applications/natural-language-inference-bert.md [hunk 9]
```diff
@@ -372,6 +370,8 @@ d2l.train_ch13(net, train_iter, test_iter, loss, trainer, num_epochs, devices,
 lr, num_epochs = 1e-4, 5
 trainer = torch.optim.Adam(net.parameters(), lr=lr)
 loss = nn.CrossEntropyLoss(reduction='none')
+# A dry run to initialize the model.
```
Comment (astonzhang) [inline chapter_natural-language-processing-applications/natural-language-inference-bert.md:373]: For now we do not include . at the end of a code comment. To save space, consider using inline comments with 2 space between the end of code and the beginning of inline comments.

File: chapter_natural-language-processing-pretraining/bert-pretraining.md [hunk 10]
```diff
@@ -175,6 +173,7 @@ def train_bert(train_iter, net, loss, vocab_size, devices, num_steps):
 ```{.python .input}
 #@tab pytorch
 def train_bert(train_iter, net, loss, vocab_size, devices, num_steps):
+    net(*next(iter(train_iter))[:4]) # A dry run to initialize the model.
```
Comment (astonzhang) [inline chapter_natural-language-processing-pretraining/bert-pretraining.md:176]: Keep 2 space right before an inline comment. Remove period at the end of the comment.

File: chapter_natural-language-processing-applications/natural-language-inference-bert.md (hunk 7 above)
Comment (astonzhang) [reply chapter_natural-language-processing-applications/natural-language-inference-bert.md]: Thanks. Resolving it now.

**Comment by astonzhang** [comment]
Since https://github.com/d2l-ai/d2l-en/pull/2112 is merged, plz also resolve the conflicts

File: chapter_natural-language-processing-applications/natural-language-inference-bert.md (hunk 9 above)
Comment (cheungdaven) [reply chapter_natural-language-processing-applications/natural-language-inference-bert.md]: it is resolved. Thanks!

File: chapter_natural-language-processing-pretraining/bert-pretraining.md (hunk 10 above)
Comment (cheungdaven) [reply chapter_natural-language-processing-pretraining/bert-pretraining.md]: It is resolved, thanks!

**Comment by cheungdaven** [comment]
> It's great to avoid hardcoding. Nice teamwork! cc @AnirudhDagar
> 
> The code looks good to me except for those open comments. Pending check: corresponding text, evaluation results Since #2112 is merged, plz also resolve the conflicts

I have resolved the conflicts, thanks!

File: chapter_computer-vision/image-augmentation.md (hunk 1 above)
Comment (AnirudhDagar) [reply chapter_computer-vision/image-augmentation.md]: It will run on multiple GPUs if we pass multiple GPUs to the argument `device_ids `. Think of it as moving the network to the GPU device 0 and since dataparallel is wrapped, with the first forward call it starts splitting into smaller batches to available devices listed in `device_ids`.
